File: lcof/cc/yong-liang-ge-zhan-shi-xian-dui-lie-lcof.cc
/*
 * @Author: FisherCloud 
 * @Date: 2020-02-27 16:13:47 
 * @Last Modified by: FisherCloud
 * @Last Modified time: 2020-02-27 16:47:50
 */
// https://leetcode-cn.com/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof/
#include <iostream>
#include <stack>

using namespace std;

class CQueue
{
public:
    CQueue()
    {
        this->stackone = new stack<int>;
        this->stacktwo = new stack<int>;
    }

    void appendTail(int value)
    {
        this->stackone->push(value);
    }

    int deleteHead()
    {
        if (this->stacktwo->empty())
        {
            if (this->stackone->empty())
                return -1;
            else
            {
                while (!this->stackone->empty())
                {
                    this->stacktwo->push(this->stackone->top());
                    this->stackone->pop();
                }
            }
        }
        int ans = this->stacktwo->top();
        this->stacktwo->pop();
        return ans;
    }

private:
    stack<int> *stackone;
    stack<int> *stacktwo;
};

/**
 * Your CQueue object will be instantiated and called as such:
 * CQueue* obj = new CQueue();
 * obj->appendTail(value);
 * int param_2 = obj->deleteHead();
 */

int main(int argc, char const *argv[])
{
    cout << "lcof-09" << endl;
    return 0;
}

File: lcof/rs/yong-liang-ge-zhan-shi-xian-dui-lie-lcof.rs
/*
 * @Author: FisherCloud 
 * @Date: 2020-02-27 16:48:36 
 * @Last Modified by: FisherCloud
 * @Last Modified time: 2020-02-27 17:22:06
 */
// https://leetcode-cn.com/problems/yong-liang-ge-zhan-shi-xian-dui-lie-lcof/

use std::collections::LinkedList;

struct CQueue {
    stackone: LinkedList<i32>, // Vec,VecDeque,LinkedList均可，但测试过后VecDeque效率更高
    stacktwo: LinkedList<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl CQueue {

    fn new() -> Self {
        CQueue{
            stackone: LinkedList::new(),
            stacktwo: LinkedList::new(),
        }
    }
    
    fn append_tail(&mut self, value: i32) {
        self.stackone.push_back(value);
    }
    
    fn delete_head(&mut self) -> i32 {
        if self.stacktwo.is_empty() {
            if self.stackone.is_empty() {
                return -1;
            } else {
                while !self.stackone.is_empty() {
                    self.stacktwo.push_back(self.stackone.pop_back().unwrap());
                }
            }   
        }
        return self.stacktwo.pop_back().unwrap();
    }
}

/**
 * Your CQueue object will be instantiated and called as such:
 * let obj = CQueue::new();
 * obj.append_tail(value);
 * let ret_2: i32 = obj.delete_head();
 */
